### ISSUE TITLE
Can NavigationParameters implement IDictionary<string, object> instead?

### ISSUE THREAD

**Issue by michaelnero** [PrismLibrary/Prism#62]
Is it possible to have `NavigationParameters` implement `IDictionary<string, object>` instead of just `IEnumerable<KeyValuePair<string, object>>`?

I find myself testing for the presence of a parameter using the Linq `Single` or `Any` methods instead of `TryGetValue` quite often. I can work around this by creating my own Dictionary populated with NavigationParameters, of course, but thought I'd check before I resorted to that.

**Comment by brianlagunas** [comment]
Are you suggesting NavigationParameters : IEnumerable<IDictionary<string,object>>?

**Comment by michaelnero** [comment]
Sorry, I'm suggesting `NavigationParameters : IDictionary<string, object>`

**Comment by bartlannoeye** [comment]
I've been using IDictionary as well, so you got my vote.

**Comment by brianlagunas** [comment]
Hmmm.. I haven't really looked into that.  Not sure if that would cause any breaking changes or not.  I wouldn't think it would.  I do know that would mean implementing a lot more methods, meaning more code to write.

Pull request anyone :)

**Comment by brianlagunas** [comment]
thanks for the pull request.  It's been merged.

**Comment by brianlagunas** [comment]
We actually just converted this back to the original API. It was too large of a breaking change for no real benefit.

**Comment by YehudahA** [comment]
Oh, I've updated all of my code, and now I will have problems again. 
In my opinion, using Dictionary is logical and understandable.

**Comment by brianlagunas** [comment]
There are a ton of other scenarios that weren't considered before we changed it which caused a ton of breaking changes for all of our other customers. So we had to revert back. I think a better approach would be to add a TryGetValue method to the NavigationParameters instead of changing its base type. You could even use an extension method if using Any or First it inconvenient for you.

Can you provide a code snippet on what you have to do now, and then another one of how you would like it to be?  This will help us better understand your scenario.

**Comment by YehudahA** [comment]
I added the following methods to deal with the revert. But I think the main difference is the indexer behavior. ```NavigationParameters["key"]``` return ```null``` if key not found, In contrast to Dictionary that throws a ```KeyNotFoundException```.

I think that Dictionary behavior is better. In general, I do not understand what the hell he thought whoever wrote this code for the first time. He had not heard about ```Dictionary```?! Why he preferred an ugly thing like ```IEnumerable<KeyValuePair<string, object>>```?!

Anyway, this is my helper:
```
    public static class NavigationParametersExtensions
    {
        public static bool ContainsKey(this NavigationParameters navigationParameters, string key)
        {
            var dic = new Dictionary<int,int>();
            return navigationParameters.Any(parameter => string.Compare(parameter.Key, key, StringComparison.Ordinal) == 0);
        }

        public static bool TryGetValue<TValue>(this NavigationParameters navigationParameters, string key, out TValue value)
        {
            value = default(TValue);
            if (navigationParameters.ContainsKey(key))
            {
                object obj = navigationParameters[key];
                value = ReflectionHelper.ChangeType<TValue>(obj);
                return true;
            }

            return false;
        }
```

**Comment by brianlagunas** [comment]
It is easy to not understand API decisions when you are not exposed to a wider range of application scenarios, especially those that have cross platform concerns.  As developers it is easy to only see things with tunnel vision, because we just have no idea of the other use cases that must be supported.

Thanks for the extension methods.

**Comment by briannoyes** [comment]
I agree that Dictionary would have been better if that had been how it started. But many prefer a dictionary like object that just returns null if the key is not found instead of either having to deal with an exception (with associated perf concerns) or do the slightly more complicated coding pattern of always having to ContainsKey first or using TryGetValue with out variable + bool return type.

The more important reason for the decision is that introducing this now could break thousands of apps out there using Prism 5 if they move to 6 with null reference exceptions. That argument trumps any "this is cleaner" arguments by a couple orders of magnitude.

**Comment by michaelnero** [comment]
Hi @brianlagunas and @briannoyes,

I know you said that you reverted the change to make `NavigationParameters : Dictionary<string, object>`, but I have Nuget package Prism.Forms v6.0.1 (the latest AFAIK), and NavigationParameters does inherit from `Dictionary<string, object>`.

I'm a bit confused. Are you expecting to revert this in the next release? Just wanted to make sure before I start using NavigationParameters like a real dictionary.

**Comment by bartlannoeye** [comment]
The `NavigationParameters` class is currently still platform-specific. The WPF-version has been reverted, but XF wasn't and UWP doesn't even have it. We'll have to look into this.

**Comment by michaelnero** [comment]
For what it's worth, if having `NavigationParameters` inherit from `Dictionary<string, object>` isn't causing problems anywhere else, I prefer to leave it that way. Just my two cents.

**Comment by brianlagunas** [comment]
Of course for WPF, we won't change it, but we will have to discuss if we want to try to share a single NavigationParameters class across platforms or not.   Not only do these classes exist in different namespaces for WPF and XF (doesn't exist for UWP yet), they are different implementations.  So, if we did standardize on one, it would have to be the WPF version.  Then we would have a breaking change for a namespace that doesn't make any sense on either XF or UWP, unless we moved the namespace for WPF too.  I am thinking we will keep them separate for now.  One class of duplication isn't going to hurt us, and I am not sure is worth the hassle for developers.
